ENH: sparse: speed up CSR/CSC toarray()

perimosocordiae opened the pull request:
This PR adds a new function to sparsetools: `csr_todense`, which has the same semantics as `coo_todense` in that it's actually performing `D += S`. Unlike the COO version, this function requires that the dense array is in C-contiguous order.

@pv: I've never added functions to sparsetools before, so please let me know if I've missed something.

The one backwards-incompatible change here is that I set the default `csc_matrix.toarray` order to be F-contiguous, for efficiency reasons. Users can still specify an order explicitly, but the `mat.A` shorthand will now produce column-major matrices for CSC inputs.

There weren't any benchmarks testing `toarray` timing, so I added one. Here are the changes between current master and this PR:
```
BENCHMARK                                    BEFORE      AFTER     FACTOR
sparse.Densify.time_toarray('csc', 'F')    635.97μs   474.44μs   0.74600847x
sparse.Densify.time_toarray('csr', 'C')    629.63μs   468.08μs   0.74341441x
```

Finally, once gh-7079 is merged, this change will allow a more efficient implementation of `_add_dense`.

pv commented on scipy/sparse/compressed.py:
```diff
@@ -960,8 +960,22 @@ def tocoo(self, copy=True):
     def toarray(self, order=None, out=None):
-        """See the docstring for `spmatrix.toarray`."""
-        return self.tocoo(copy=False).toarray(order=order, out=out)
+        ideal = self._swap('cf')[0]
```
I guess this is a backward compatibility break?
Maybe we can do without it?

pv commented:
LGTM, although I would leave the backward compat break out.

perimosocordiae commented:
I rewrote some history and split this up a bit.

CSR/CSC + dense now avoids conversion to COO, enabled by gh-7079.

The backwards-incompatibility is now isolated to commit 726239eb0fe71ebf10deb91bee753b4138f11f9a. My personal opinion is that this won't adversely affect much user code, and that the upcoming 1.0 release seems like as good a time as we'll get for changes like this. Also, if the previous behavior is needed, users can always specify an explicit order in `.toarray()`.

That said, I'll defer to the consensus on whether it's worth breaking compatibility for the speedup here.
